My nc news 7

kamiviolet opened the pull request:
1. update endpoints.json
2. add new method to endpoint /api/articles/:article_id/comments
3. add test suite
- 2 tests for status 201
- 5 tests for status 400 (as it is a POST method, i decided to use 400 even if the numeric article_id is not found as it is not to GET the article but a REQUEST)

updated..

kamiviolet committed:
add POST /api/articles/:article_id/comments and related utils, test suite

kamiviolet committed:
update endpoints.json for POST /api/articles/:article_id/comment

kamiviolet committed:
declare variable to store the posting object in test suite

kamiviolet committed:
add 1 more test for the newly created endpoint

kamiviolet committed:
Merge updated branch 'main' after approval into earlier worked my-nc-news-7

kamiviolet committed:
add 1 more test to check if there is request body

yalicee commented on __tests__/app.test.js:
```diff
@@ -218,6 +221,7 @@ describe('/api/articles/:article_id/comments', () => {
         })
     })
+
     it('GET - status 404 - invalid numeric article_id will respond with not found.', () => {
```
technically this is a valid numeric article but just non existent

yalicee commented on __tests__/app.test.js:
```diff
@@ -227,4 +231,105 @@ describe('/api/articles/:article_id/comments', () => {
+                expect(rows.length).toBe(1);
+                expect(rows[0].author).toBe(example.username);
+                expect(rows[0].body).toBe(example.body);
+                expect(rows[0].article_id).toBe(article_id);
```
SQL is not needed in testing. This is similar to the test you have above as you return back the comment object anyway. Please remove

yalicee commented on __tests__/app.test.js:
```diff
@@ -227,4 +231,105 @@ describe('/api/articles/:article_id/comments', () => {
+        })
+    })
+
+    it('POST - status 400 - no request body is sent.', () => {
+        return request(app)
+            .post('/api/articles/2/comments')
+            .send({})
+            .expect(400)
+            .then(({body}) => {
+                const {message} = body;
+                expect(message).toBe('It seems you forget to send the request...')
+        })
+    })
```
This test could be considered redundant due to test above

yalicee commented on __tests__/app.test.js:
```diff
@@ -227,4 +231,105 @@ describe('/api/articles/:article_id/comments', () => {
+        })
+    })
+
+    it('POST - status 400 - the article_id is not existing.', () => {
+        return request(app)
+            .post('/api/articles/99999/comments')
+            .send({username: 'XXX', body: 'XXXX'})
+            .expect(400)
+            .then(({body}) => {
+                const {message} = body;
+                expect(message).toBe('Apology. Non-existing article_id.')
+        })
+    })
```
This is a 404 not found test

yalicee commented on __tests__/app.test.js:
```diff
@@ -227,4 +231,105 @@ describe('/api/articles/:article_id/comments', () => {
+    })
+
+
+    it('POST - status 400 - the username is not existing.', () => {
+        return request(app)
+            .post('/api/articles/2/comments')
+            .send({username: 'XXX', body: 'XXXX'})
+            .expect(400)
+            .then(({body}) => {
+                const {message} = body;
+                expect(message).toBe('Apology. Non-existing username, please sign up to leave your comment.')
+        })
```
Same for this one

yalicee commented on controllers/comments.controller.js:
```diff
@@ -12,4 +12,25 @@ const getCommentsByArticleId = (req, res, next) => {
+        next(err);
+    } else if (Number.isNaN(+article_id) !== false) {
+        const err = {status: 400, message: 'Invalid article_id.'}
+        next(err);
```
Have a think about if you can move this functionality somewhere else? Perhaps in your error handlers?

yalicee commented on db/seeds/utils.js:
```diff
@@ -32,7 +32,21 @@ exports.validateExistingId = (id) => {
+    })
+}
+
+exports.validateExistingUser = (username) => {
+  return db
+    .query(`
+      SELECT * 
+      FROM users
+      WHERE username in ($1)
+    `, [username])
+    .then(({rows}) => {
+      if (rows.length === 0) {
+        return {status: 404, message: 'The username is currently not found.'}
```
author in the comments table is a foreign key which relates to username in the users table. A psql error will be thrown if this is violated. It is not necessary to separately check if the user exists. same for article_id

yalicee commented on models/comments.model.js:
```diff
@@ -1,16 +1,45 @@
+}
+
+exports.createNewCommentByArticleId = (id, comment) => {
+    return validateExistingId(id)
+        .then((msg) => {
+            if (msg !== undefined) {
+                return Promise.reject({status: 400, message: 'Apology. Non-existing article_id.'})
+            }
+        })
+        .then(() => validateExistingUser(comment.username))
```
Again no need to check if they exist

yalicee reviewed (changes requested):
Other things to consider besides my comments:
-article_id is also a foreign key so again, a separate check that the article exists should not be necessary as a psql error will be thrown.
-a psql error will also be thrown if article_id is not a number.

yalicee reviewed (changes requested):
Other things to consider besides my comments:
-article_id is also a foreign key so again, a separate check that the article exists should not be necessary as a psql error will be thrown.
-a psql error will also be thrown if article_id is not a number.

kamiviolet commented:
> Other things to consider besides my comments: -article_id is also a foreign key so again, a separate check that the article exists should not be necessary as a psql error will be thrown. -a psql error will also be thrown if article_id is not a number.

@yalicee Sorry i am stuck... i added the extra util function to check if its a valid foreign key because if not doing so, my terminal is hanging from Jest 500s timeout... how should i organise to be able to achieve so? T_T

I tried to add extra catch appended to the db.query in model file.. (not recommended to have catch everywhere but thats the only way to avoid hanging and timeout) but then the controller can log the psql error but not able to get to my error handler in app.js... how so? hm...

Otherwise I have adjusted the status and statements to be more accurately reflecting to the situation accordingly

kamiviolet committed:
add to error handler based on various psql error codes

kamiviolet committed:
delete unnecessary util function for validating user (foreign key)

kamiviolet committed:
add script jest cli to detect where to have test leak

kamiviolet committed:
change status and message related to sad paths for new method

kamiviolet committed:
erase unnecessary logics in controller and model

yalicee commented on __tests__/app.test.js:
```diff
@@ -218,13 +221,82 @@ describe('/api/articles/:article_id/comments', () => {
+    it('POST - status 404 - the article_id is not existing.', () => {
+        return request(app)
+            .post('/api/articles/99999/comments')
+            .send({username: 'XXX', body: 'XXXX'})
```
Please change this to be a valid comment object so it's less confusing what error it is for

yalicee commented on app.js:
```diff
@@ -21,20 +21,32 @@ app.route('/api/articles/:article_id')
     }
+
+    res.status(400).send({message: 'Bad request.'});
 })
```
There is an opportunity to refactor this later on

yalicee commented on controllers/comments.controller.js:
```diff
@@ -12,4 +12,17 @@ const getCommentsByArticleId = (req, res, next) => {
+
+    return createNewCommentByArticleId(article_id, newComment)
+        .then((comment) => {
+            res.status(201).send({comment: comment[0]})
```
would prefer you to index this array in the model instead of the controller

kamiviolet committed:
separate all error handlers to ./controller

kamiviolet committed:
refactor psql error handler

kamiviolet committed:
change the comment body for testing POST new comment
